### PR TITLE
Issue 4822: Updated to MSVC 2019

### PR DESCRIPTION
**Pull request by JCash**
Main reason was to update to new package format

### CODE REVIEW

File: build_tools/waf_dynamo.py [hunk 1]
```diff
@@ -1284,120 +1286,49 @@ def create_clang_wrapper(conf, exe):
     os.chmod(clang_wrapper_path, stat.S_IRUSR | stat.S_IWUSR | stat.S_IXUSR)
     return clang_wrapper_path
 
-# Capture vcvarsall.bat environment variables for MSVC binaries, INCLUDE and LIB dir
```
Comment (JCash) [inline build_tools/waf_dynamo.py:1287]: Removed old unused stuff

File: build_tools/waf_dynamo.py [hunk 2]
```diff
@@ -1284,120 +1286,49 @@ def create_clang_wrapper(conf, exe):
...
     ucrt_dirs = [ x for x in os.listdir(os.path.join(windowskitsdir,'10','Include'))]
     ucrt_dirs = [ x for x in ucrt_dirs if x.startswith('10.0')]
     ucrt_dirs.sort(key=lambda x: int((x.split('.'))[2]))
     ucrt_version = ucrt_dirs[-1]
     if not ucrt_version.startswith('10.0'):
         conf.fatal("Unable to determine ucrt version: '%s'" % ucrt_version)
 
-    includes = [os.path.join(msvcdir,'VC','include'),
-                os.path.join(msvcdir,'VC','atlmfc','include'),
+    msvc_version = [x for x in os.listdir(os.path.join(msvcdir,'VC','Tools','MSVC'))]
```
Comment (JCash) [inline build_tools/waf_dynamo.py:1308]: Finds the version automatically (assuming they stop moving folders around)

File: scripts/build.py [hunk 3]
```diff
@@ -65,9 +65,11 @@
 PACKAGES_IOS_SIMULATOR_SDK="iPhoneSimulator13.1.sdk"
 PACKAGES_MACOS_SDK="MacOSX10.15.sdk"
 PACKAGES_XCODE_TOOLCHAIN="XcodeDefault11.1.xctoolchain"
-PACKAGES_WIN32_TOOLCHAIN="Microsoft-Visual-Studio-14-0"
+WINDOWS_SDK_10_VERSION="10.0.18362.0"
+WINDOWS_MSVC_2019_VERSION="14.25.28610"
+PACKAGES_WIN32_TOOLCHAIN="Microsoft-Visual-Studio-2019-{0}".format(WINDOWS_MSVC_2019_VERSION)
```
Comment (JCash) [inline scripts/build.py:70]: Now uses the same packages as the extender server.

File: scripts/build.py [hunk 4]
```diff
@@ -503,7 +505,7 @@ def download_sdk(url, targetfolder, strip_components=1):
             win32_sdk_folder = join(self.ext, 'SDKs', 'Win32')
             download_sdk( '%s/%s.tar.gz' % (self.package_path, PACKAGES_WIN32_SDK_8), join(win32_sdk_folder, 'WindowsKits', '8.1') )
             download_sdk( '%s/%s.tar.gz' % (self.package_path, PACKAGES_WIN32_SDK_10), join(win32_sdk_folder, 'WindowsKits', '10') )
-            download_sdk( '%s/%s.tar.gz' % (self.package_path, PACKAGES_WIN32_TOOLCHAIN), join(win32_sdk_folder, 'MicrosoftVisualStudio14.0') )
+            download_sdk( '%s/%s.tar.gz' % (self.package_path, PACKAGES_WIN32_TOOLCHAIN), join(win32_sdk_folder, 'MicrosoftVisualStudio14.0'), strip_components=0 )
```
Comment (JCash) [inline scripts/build.py:508]: `strip_components = 0` to avoid removing the first part of the hierarchy.

File: build_tools/waf_dynamo.py [hunk 5]
```diff
@@ -1284,122 +1286,45 @@ def create_clang_wrapper(conf, exe):
...
+    msvc_version = [x for x in msvc_version if x.startswith('14.')]
+    msvc_version.sort(key=lambda x: map(int, x.split('.')))
+    msvc_version = msvc_version[-1]
+    if not msvc_version.startswith('14.'):
+        conf.fatal("Unable to determine msvc version: '%s'" % msvc_version)
+
+    msvc_path = (os.path.join(msvcdir,'VC', 'Tools', 'MSVC', msvc_version, 'bin', 'Host'+arch, arch),
+                os.path.join(windowskitsdir, '8.1', 'bin', arch))
```
Comment (eangelin) [inline build_tools/waf_dynamo.py:1315]: Shouldn't all of these 8.1's be 10's?

File: scripts/build.py [hunk 6]
```diff
@@ -65,9 +65,11 @@
 PACKAGES_IOS_SIMULATOR_SDK="iPhoneSimulator13.1.sdk"
 PACKAGES_MACOS_SDK="MacOSX10.15.sdk"
 PACKAGES_XCODE_TOOLCHAIN="XcodeDefault11.1.xctoolchain"
-PACKAGES_WIN32_TOOLCHAIN="Microsoft-Visual-Studio-14-0"
+WINDOWS_SDK_10_VERSION="10.0.18362.0"
+WINDOWS_MSVC_2019_VERSION="14.25.28610"
+PACKAGES_WIN32_TOOLCHAIN="Microsoft-Visual-Studio-2019-{0}".format(WINDOWS_MSVC_2019_VERSION)
 PACKAGES_WIN32_SDK_8="WindowsKits-8.1"
```
Comment (eangelin) [inline scripts/build.py:71]: Shouldn't we remove 8.1 entirely?

File: scripts/build.py [hunk 7]
```diff
@@ -503,7 +505,7 @@ def download_sdk(url, targetfolder, strip_components=1):
             win32_sdk_folder = join(self.ext, 'SDKs', 'Win32')
             download_sdk( '%s/%s.tar.gz' % (self.package_path, PACKAGES_WIN32_SDK_8), join(win32_sdk_folder, 'WindowsKits', '8.1') )
```
Comment (eangelin) [inline scripts/build.py:506]: ... and this

File: scripts/package/package_win32_sdk.sh [hunk 8]
```diff
@@ -60,12 +60,14 @@ if [ ! -e "${TARGET_PATH}/${PACKAGES_WIN32_TOOLCHAIN}" ]; then
 	TMP=MicrosoftVisualStudio2019
```
Comment (eangelin) [inline scripts/package/package_win32_sdk.sh:61]: Not related to this PR, but VC_PATH above assumes Community edition, we should mention in a readme you need to change this if you have Professional. Or even better detect it :)

File: scripts/package/package_win32_sdk.sh [hunk 9]
```diff
@@ -60,12 +60,14 @@ if [ ! -e "${TARGET_PATH}/${PACKAGES_WIN32_TOOLCHAIN}" ]; then
...
 	mkdir -p $TMP/VC/Tools/MSVC/$MSVC_VERSION/include
 	mkdir -p $TMP/VC/Tools/MSVC/$MSVC_VERSION/lib/x64
 	mkdir -p $TMP/VC/Tools/MSVC/$MSVC_VERSION/lib/x86
 	mkdir -p $TMP/VC/Tools/MSVC/$MSVC_VERSION/atlmfc
 
 	cp -r -v "$VC_PATH/VC/Tools/MSVC/$MSVC_VERSION/bin/Hostx64/x64" "$TMP/VC/Tools/MSVC/$MSVC_VERSION/bin/Hostx64"
+	cp -r -v "$VC_PATH/VC/Tools/MSVC/$MSVC_VERSION/bin/Hostx64/x86" "$TMP/VC/Tools/MSVC/$MSVC_VERSION/bin/Hostx86"
 	cp -r -v "$VC_PATH/VC/Tools/MSVC/$MSVC_VERSION/include" "$TMP/VC/Tools/MSVC/$MSVC_VERSION"
 	cp -r -v "$VC_PATH/VC/Tools/MSVC/$MSVC_VERSION/lib/x64" "$TMP/VC/Tools/MSVC/$MSVC_VERSION/lib"
 	cp -r -v "$VC_PATH/VC/Tools/MSVC/$MSVC_VERSION/lib/x86" "$TMP/VC/Tools/MSVC/$MSVC_VERSION/lib"
```
Comment (eangelin) [inline scripts/package/package_win32_sdk.sh:73]: Also noticed this script doesn't remove the TMP dir 'MicrosoftVisualStudio2019' after tar'ing

**Pull request merged**